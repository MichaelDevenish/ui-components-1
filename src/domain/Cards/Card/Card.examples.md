#### Expandable Card

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

initialState = { card1: false, card2: false, allCards: false };

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';

<div>
  <Button
    type='neutral'
    onClick={() => setState({ card1: !state.allCards, card2: !state.allCards, allCards: !state.allCards})}>
    {state.allCards ? 'Collapse All' : 'Expand All'}
  </Button>
  <Card
    margins={{
              top:12,
              bottom: 4
            }}
    isExpanded= {state.card1}
    onCardToggle= {() => setState({ card1: !state.card1})}
    dropdownSections= {[
        {
          text: 'homepage',
          href: 'https://intellihr.github.io/ui-components/',
          stopPropagation: true
        },
        {
          text: 'Delete',
          onClick: (event) => { alert('Delete action for the card') },
          sectionType: 'alert',
          stopPropagation: true
        }
      ]}
    mainContent= {
      <GridLayout
        gutterPaddingX={Variables.Spacing.sSmall}
        gutterPaddingY={Variables.Spacing.sSmall}
        cells={[
          {
            size: 7,
            content: text
          },
          {
            size: 5,
            content: text
          }
        ]}
      />
    }
    extraContent= {
      <GridLayout
        gutterPaddingX={Variables.Spacing.sSmall}
        gutterPaddingY={Variables.Spacing.sSmall}
        cells={[
          {
            size: 12,
            content: text
          },
          {
            size: 12,
            content: text
          }
        ]}
      />
    }
  />
  <Card
      color='orange'
      isExpanded= {state.card2}
      onCardToggle= {() => setState({ card2: !state.card2})}
      dropdownSections= {[
          {
            text: 'google',
            href: 'https://www.google.com.au',
            stopPropagation: true
          },
          {
            text: 'Delete',
            onClick: (event) => { alert('Delete action for the card') },
            sectionType: 'alert',
            stopPropagation: true
          }
        ]}
      mainContent= {
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 7,
              content: text
            },
            {
              size: 5,
              content: text
            }
          ]}
        />
      }
      extraContent= {
        <GridLayout
          gutterPaddingX={Variables.Spacing.sSmall}
          gutterPaddingY={Variables.Spacing.sSmall}
          cells={[
            {
              size: 12,
              content: text
            },
            {
              size: 12,
              content: text
            }
          ]}
        />
      }
    />
</div>
```

#### Expanded Card without action

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

initialState = { cardWithOutAction: false };

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';

<Card
    isExpanded= {state.cardWithOutAction}
    onCardToggle= {() => setState({ cardWithOutAction: !state.cardWithOutAction})}
    mainContent= {
      <GridLayout
        gutterPaddingX={Variables.Spacing.sSmall}
        gutterPaddingY={Variables.Spacing.sSmall}
        cells={[
          {
            size: 7,
            content: text
          },
          {
            size: 5,
            content: text
          }
        ]}
      />
    }
    extraContent= {
      <GridLayout
        gutterPaddingX={Variables.Spacing.sSmall}
        gutterPaddingY={Variables.Spacing.sSmall}
        cells={[
          {
            size: 12,
            content: text
          },
          {
            size: 12,
            content: text
          }
        ]}
      />
    }
  />
```

#### Card with actions

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';

<Card
    isExpanded= {state.cardWithOutExpand}
    dropdownSections= {[
              {
                text: 'google',
                href: 'https://www.google.com.au',
                stopPropagation: true
              },
              {
                text: 'Delete',
                onClick: (event) => { alert('Delete action for the card') },
                sectionType: 'alert',
                stopPropagation: true
              }
            ]}
    mainContent= {
      <GridLayout
        gutterPaddingX={Variables.Spacing.sSmall}
        gutterPaddingY={Variables.Spacing.sSmall}
        cells={[
          {
            size: 7,
            content: text
          },
          {
            size: 5,
            content: text
          }
        ]}
      />
    }
  />
```

#### Simple card

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';

<div>
  <Card
    margins={{bottom: 8}}
    mainContent= {
      <GridLayout
        gutterPaddingX={Variables.Spacing.sSmall}
        gutterPaddingY={Variables.Spacing.sSmall}
        cells={[
          {
            size: 7,
            content: text
          },
          {
            size: 5,
            content: text
          }
        ]}
      />
    }
  />
</div>
```

#### Card will manage its own state if *isExpanded* is not passed in.

```jsx
import { Button } from '@Domain/Buttons';
import { GridLayout } from '@Domain/Layouts';
import { Variables } from '@Common';

const text = 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Curabitur nec libero et libero molestie eleifend. Donec dignissim vel erat eu cursus.';

<Card
  margins={{
            top:12,
            bottom: 4
          }}
  dropdownSections= {[
      {
        text: 'homepage',
        href: 'https://intellihr.github.io/ui-components/',
        stopPropagation: true
      },
      {
        text: 'Delete',
        onClick: (event) => { alert('Delete action for the card') },
        sectionType: 'alert',
        stopPropagation: true
      }
    ]}
  mainContent= {
    <GridLayout
      gutterPaddingX={Variables.Spacing.sSmall}
      gutterPaddingY={Variables.Spacing.sSmall}
      cells={[
        {
          size: 7,
          content: text
        },
        {
          size: 5,
          content: text
        }
      ]}
    />
  }
  extraContent= {
    <GridLayout
      gutterPaddingX={Variables.Spacing.sSmall}
      gutterPaddingY={Variables.Spacing.sSmall}
      cells={[
        {
          size: 12,
          content: text
        },
        {
          size: 12,
          content: text
        }
      ]}
    />
  }
/>
```
