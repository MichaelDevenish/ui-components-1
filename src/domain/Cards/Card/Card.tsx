import React from 'react'

import { Props } from '../../../common'
import { FontAwesomeIcon } from '../../Icons/FontAwesomeIcon'
import { DropdownMenu, IDropdownMenuToggleComponentProps } from '../../Popovers/DropdownMenu'
import { ISectionProps } from '../../Popovers/DropdownMenu/subcomponents/Section'
import {
  ChevronIconWrapper,
  StyledPrimaryContent
} from '../services/style'
import {
  StyledActionButton,
  StyledBodyContents,
  StyledCard,
  StyledCardHeader,
  StyledCardToggleButton,
  StyledExtraContent
} from './style'

type CardColors = 'neutral' | 'orange' | 'red' | 'grey'

interface ICardProps {
  /** the component that shows even in collapse mode */
  mainContent: JSX.Element
  /** the component that shows in expand mode only */
  extraContent?: JSX.Element
  /** If true, the card is in expand style */
  isExpanded?: boolean
  /** dropwon sections to show in the cards action button dropdown */
  dropdownSections?: ISectionProps[]
  /** callback when user toggle card */
  onCardToggle?: () => void
  /** The margins around the component */
  margins?: Props.IMargins
  /** The data-component-context */
  componentContext?: string
  /** If the card should have a hover style */
  isHoverable?: boolean
  /** Background color of the card */
  color?: CardColors
}

interface ICardState {
  isActionButtonHover: boolean
  isExpanded: boolean
}

class Card extends React.PureComponent<ICardProps, ICardState> {
  public static defaultProps: Partial<ICardProps> = {
    isHoverable: false,
    color: 'neutral'
  }

  public state: ICardState = {
    isActionButtonHover: false,
    isExpanded: false
  }

  public render (): JSX.Element {
    const {
      mainContent,
      extraContent,
      componentContext,
      margins,
      onCardToggle,
      isHoverable,
      color
    } = this.props

    return (
      <StyledCard
        margins={margins}
        color={color!}
        data-component-type={Props.ComponentType.Card}
        data-component-context={componentContext}
      >
        <StyledCardHeader
          onClick={this.handleCardToggle}
          isExpanded={this.isExpanded && !!extraContent}
          hasHoverStyle={!this.state.isActionButtonHover && (!!extraContent || !!onCardToggle || isHoverable!)}
          color={color!}
        >
          <StyledPrimaryContent>{mainContent}</StyledPrimaryContent>
          {this.actionButtonDropdownMenu}
          {this.toggleButton}
        </StyledCardHeader>
        {extraContent && (
          <StyledExtraContent isExpanded={this.isExpanded} color={color!}>
            <StyledBodyContents>
              {extraContent}
            </StyledBodyContents>
          </StyledExtraContent>
        )}
      </StyledCard>
    )
  }

  private handleCardToggle = () => {
    this.setState((state) => ({ isExpanded: !state.isExpanded }))

    if (this.props.onCardToggle) {
      this.props.onCardToggle()
    }
  }

  private get isExpanded () {
    if (this.props.isExpanded === undefined) {
      return this.state.isExpanded
    }

    return this.props.isExpanded
  }

  private get actionButtonDropdownMenu (): JSX.Element | null {
    const {
      dropdownSections
    } = this.props

    if (dropdownSections) {
      return (
        <DropdownMenu
          sections={dropdownSections}
          toggleComponent={this.actionButton}
        />
      )
    }

    return null
  }

  private actionButton = ({ toggleMenu, toggleComponentRef, ariaProps }: IDropdownMenuToggleComponentProps) => (
    <StyledActionButton
      onClick={this.handleActionButtonClick(toggleMenu)}
      onMouseOver={this.handleActionButtonMouseOver}
      onMouseOut={this.handleActionButtonMouseOut}
      ref={toggleComponentRef}
      hasRightMargin={!!this.props.extraContent}
      color={this.props.color!}
      {...ariaProps}
    >
      <FontAwesomeIcon type='solid' icon='ellipsis-v' />
    </StyledActionButton>
  )

  private get toggleButton (): JSX.Element | undefined {
    const {
      extraContent,
      color
    } = this.props

    if (extraContent) {
      return (
        <StyledCardToggleButton isExpanded={this.isExpanded} hasParentHoverStyle={!this.state.isActionButtonHover} color={color!}>
          <ChevronIconWrapper>
            <FontAwesomeIcon type='solid' icon='chevron-down' />
          </ChevronIconWrapper>
        </StyledCardToggleButton>
      )
    }
  }

  private handleActionButtonClick = (toggleMenu: () => void) => (event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation()
    toggleMenu()
  }

  private handleActionButtonMouseOver = () => {
    this.setState({ isActionButtonHover: true })
  }

  private handleActionButtonMouseOut = () => {
    this.setState({ isActionButtonHover: false })
  }
}

export {
  Card,
  CardColors
}
