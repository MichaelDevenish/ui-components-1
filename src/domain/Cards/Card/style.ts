import styled, { css } from 'styled-components'

import { Props, Utils, Variables } from '../../../common'
import { styleForMargins } from '../../Spacers/services/margins'
import { colorOptions } from '../services/colors'
import { cardButtonStyle, cardCollpaseAnimation, cardExpandAnimation } from '../services/style'
import { CardColors } from './Card'

import {
  StyledFlexContent
} from '../services/style'

interface IStyledExtraContentProps {
  isExpanded: boolean
  color: CardColors
}

interface IStyledCardProps {
  margins?: Props.IMargins
  color: CardColors
}

interface IStyledCardToggleButtonProps {
  isExpanded: boolean,
  hasParentHoverStyle: boolean
  color: CardColors
}

interface IStyleActionButtonProps {
  hasRightMargin: boolean
  color: CardColors
}

interface IStyledCardHeader {
  isExpanded: boolean
  hasHoverStyle: boolean
  color: CardColors
}

const StyledCard = styled.div`
  margin: 0;

  ${(props: IStyledCardProps) => styleForMargins(props.margins)}

  background-color: ${(props: IStyledCardProps) => colorOptions[props.color].background};
`
const StyledCardHeader = styled(StyledFlexContent)`
  padding: ${Variables.Spacing.sMedium}px;
  border: 1px solid ${(props: IStyledCardHeader) => colorOptions[props.color].border};
  border-radius: ${Variables.Style.borderRadius}px;

  ${(props: IStyledCardHeader) => props.hasHoverStyle && css`
      cursor: pointer;

      &:hover {
        background-color: ${colorOptions[props.color].hoverBackground};
        transition: background-color .25s ease-out;
      }
  `}

  ${(props: IStyledCardHeader) => props.isExpanded && css`
      border-radius: ${Variables.Style.borderRadius}px ${Variables.Style.borderRadius}px 0 0;
  `}
`

const StyledCardToggleButton = styled.button`
  ${cardButtonStyle};
  transition: all .25s ease-out;

  ${(props: IStyledCardToggleButtonProps) => props.isExpanded && css`
      transform: rotate(180deg);
      transition: all .25s ease-out;
  `}

  ${(props: IStyledCardToggleButtonProps) => props.hasParentHoverStyle && css`
    ${StyledCard}:hover & {
      color: ${Variables.Color.n800};
      background-color: ${colorOptions[props.color].hoverButtonBackground};
    }
  `}

  ${Utils.mediaQueryBetweenSizes({ maxPx: Variables.Breakpoint.breakpointTablet })} {
    display: none;
  }
`

const StyledActionButton = styled.button<IStyleActionButtonProps>`
  ${cardButtonStyle};

  &:hover {
    background-color: ${(props: IStyleActionButtonProps) => colorOptions[props.color].hoverButtonBackground};
  }

  ${(props: IStyleActionButtonProps) => props.hasRightMargin && css`
      margin-right: ${Variables.Spacing.sMedium}px;
  `}

  ${Utils.mediaQueryBetweenSizes({ maxPx: Variables.Breakpoint.breakpointTablet })} {
    margin-right: 0
  }
`

const StyledExtraContent = styled.div`
  padding: 0;
  height: auto;
  max-height: 0;
  overflow: hidden;
  transition: max-height .5s;
  ${cardCollpaseAnimation}
  border: 1px solid ${Variables.Color.n250};
  border-top: 0;
  border-radius: 0 0 ${Variables.Style.borderRadius}px ${Variables.Style.borderRadius}px;
  background-color: ${(props: IStyledExtraContentProps) => colorOptions[props.color].extraContentBackground};

  ${(props: IStyledExtraContentProps) => props.isExpanded && css`
      max-height: 999px; // Magic number to keep animation working when expanding
      ${cardExpandAnimation}
  `}
`

const StyledBodyContents = styled.div`
  padding: ${Variables.Spacing.sMedium}px;
`

export {
  StyledBodyContents,
  StyledCard,
  StyledCardHeader,
  StyledExtraContent,
  StyledActionButton,
  StyledCardToggleButton
}
