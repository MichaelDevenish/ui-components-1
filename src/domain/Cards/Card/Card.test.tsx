import { shallow } from 'enzyme'
import React from 'react'

import { Card } from './Card'

const dummyFunction = () => (alert('dummy'))

describe('<CardList />', () => {
  it('should render a simple expanded card with expanded component and actions ', () => {
    const wrapper = shallow(
      <Card
        collapsedComponent={<div>test2</div>}
        expandedComponent={<div>test2</div>}
        onCardToggle={dummyFunction}
        name='card1'
        isExpanded={false}
        actions={[
          {
            text: 'action1',
            href: 'https://www.google.com.au'
          },
          {
            text: 'action2',
            href: 'https://www.google.com.au'
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a expanded card with expanded component only', () => {
    const wrapper = shallow(
      <Card
        collapsedComponent={<div>test2</div>}
        expandedComponent={<div>test2</div>}
        onCardToggle={dummyFunction}
        name='card1'
        isExpanded={false}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a expanded card with actions only', () => {
    const wrapper = shallow(
      <Card
        collapsedComponent={<div>test2</div>}
        onCardToggle={dummyFunction}
        name='card1'
        isExpanded={false}
        actions={[
          {
            text: 'action1',
            href: 'https://www.google.com.au'
          },
          {
            text: 'action2',
            href: 'https://www.google.com.au'
          }
        ]}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a expanded card without expand component and actions', () => {
    const wrapper = shallow(
      <Card
        collapsedComponent={<div>test2</div>}
        onCardToggle={dummyFunction}
        name='card1'
        isExpanded={false}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
