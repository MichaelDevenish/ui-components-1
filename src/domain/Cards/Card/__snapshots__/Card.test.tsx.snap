// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Card /> should render a ard with dropdownSections only 1`] = `
<styled.div
  data-component-type="card"
  hasHoverStyle={true}
  onClick={[Function]}
>
  <styled.div>
    <styled.div>
      <div>
        test2
      </div>
    </styled.div>
    <DropdownMenu
      hasInitialFocus={false}
      sections={
        Array [
          Object {
            "href": "https://www.google.com.au",
            "text": "action1",
          },
          Object {
            "href": "https://www.google.com.au",
            "text": "action2",
          },
        ]
      }
      toggleComponent={[Function]}
    />
  </styled.div>
  <styled.div
    isExpanded={false}
  />
</styled.div>
`;

exports[`<Card /> should render a card with main content and dropdownSections  1`] = `
<styled.div
  data-component-type="card"
  hasHoverStyle={true}
  onClick={[Function]}
>
  <styled.div>
    <styled.div>
      <div>
        test2
      </div>
    </styled.div>
    <DropdownMenu
      hasInitialFocus={false}
      sections={
        Array [
          Object {
            "href": "https://www.google.com.au",
            "text": "action1",
          },
          Object {
            "href": "https://www.google.com.au",
            "text": "action2",
          },
        ]
      }
      toggleComponent={[Function]}
    />
    <styled.button
      hasParentHoverStyle={true}
      isExpanded={false}
    >
      <styled.span>
        <FontAwesomeIcon
          type="chevron-down"
        />
      </styled.span>
    </styled.button>
  </styled.div>
  <styled.div
    isExpanded={false}
  >
    <div>
      test2
    </div>
  </styled.div>
</styled.div>
`;

exports[`<Card /> should render a card with main content only 1`] = `
<styled.div
  data-component-type="card"
  hasHoverStyle={true}
  onClick={[Function]}
>
  <styled.div>
    <styled.div>
      <div>
        test2
      </div>
    </styled.div>
    <styled.button
      hasParentHoverStyle={true}
      isExpanded={false}
    >
      <styled.span>
        <FontAwesomeIcon
          type="chevron-down"
        />
      </styled.span>
    </styled.button>
  </styled.div>
  <styled.div
    isExpanded={false}
  >
    <div>
      test2
    </div>
  </styled.div>
</styled.div>
`;

exports[`<Card /> should render a card without main content and dropdownSections 1`] = `
<styled.div
  data-component-type="card"
  hasHoverStyle={true}
  onClick={[Function]}
>
  <styled.div>
    <styled.div>
      <div>
        test2
      </div>
    </styled.div>
  </styled.div>
  <styled.div
    isExpanded={false}
  />
</styled.div>
`;
